src/optimizer/join_elimination.cpp

```diff
@@ -0,0 +1,148 @@
+	case LogicalOperatorType::LOGICAL_ASOF_JOIN:
+	case LogicalOperatorType::LOGICAL_DELIM_JOIN:
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN: {
+		auto &comp_join = op->Cast<LogicalComparisonJoin>();
+		idx_t inner_child_idx;
+		auto join_type = comp_join.join_type;
+		// We only handle the left outer join and right outer join now.
+		if (IsLeftOuterJoin(join_type)) {
+			inner_child_idx = 1;
+		} else if (IsRightOuterJoin(join_type)) {
+			inner_child_idx = 0;
+		} else {
+			break;
+		}
+
+		// Check whether there are used columns for the parent node are came from the inner side.
+		bool all_matched = true;
+		auto &inner_child = comp_join.children[inner_child_idx];
+		const auto all_inner_bindings = inner_child->GetColumnBindings();
+		for (idx_t col_idx = 0; col_idx < all_inner_bindings.size(); col_idx++) {
+			if (column_references.find(all_inner_bindings[col_idx]) != column_references.end()) {
+				all_matched = false;
+				break;
+			}
+		}
+		if (!all_matched) {
+			break;
+		}
+
+		// Check whether all of the unique columns are came from the outer side.
+		auto &outer_child = comp_join.children[1 ^ inner_child_idx];
+		column_binding_set_t all_outer_bindings;
+		for (const auto &binding : outer_child->GetColumnBindings()) {
+			all_outer_bindings.insert(binding);
+		}
+		for (auto &binding : unique_column_references) {
+			if (all_outer_bindings.find(binding) == all_outer_bindings.end()) {
+				all_matched = false;
+				break;
+			}
+		}
+
+		if (all_matched) {
+			op = std::move(outer_child);
+		}
+
+		break;
+	}
+	case LogicalOperatorType::LOGICAL_AGGREGATE_AND_GROUP_BY: {
+		auto &aggr = op->Cast<LogicalAggregate>();
+		for (idx_t col_idx = 0; col_idx < aggr.expressions.size(); col_idx++) {
+			if (aggr.expressions[col_idx]->GetExpressionClass() != ExpressionClass::BOUND_AGGREGATE) {
+				continue;
+			}
+			auto &aggr_expr = aggr.expressions[col_idx]->Cast<BoundAggregateExpression>();
+			// Collect the unique columns.
+			bool has_unique_column = false;
+			if (aggr_expr.IsDistinct()) {
+				has_unique_column = true;
+			} else {
+				auto agg_expr_function_name = aggr_expr.function.name;
+				if (agg_expr_function_name == "max" || agg_expr_function_name == "min") {
+					has_unique_column = true;
+				}
+			}
+
+			if (has_unique_column) {
+				for (auto &child : aggr_expr.children) {
+					if (child->expression_class == ExpressionClass::BOUND_COLUMN_REF) {
+						unique_column_references.insert(child->Cast<BoundColumnRefExpression>().binding);
+					}
+				}
+			}
+		}
+		break;
+	}
+	case LogicalOperatorType::LOGICAL_DISTINCT: {
+		auto &distinct = op->Cast<LogicalDistinct>();
+		for (idx_t col_idx = 0; col_idx < distinct.distinct_targets.size(); col_idx++) {
+			if (distinct.distinct_targets[col_idx]->GetExpressionClass() != ExpressionClass::BOUND_COLUMN_REF) {
+				continue;
+			}
+			unique_column_references.insert(
+			    distinct.distinct_targets[col_idx]->Cast<BoundColumnRefExpression>().binding);
```

Comment:
I don't think this is correct since the bindings in the `DISTINCT` refer to a different node than the bindings in a following join. The table index in a `ColumnBinding` refer to an operator - in a distinct those will refer to a *child* of the distinct node. In the join the table binding will refer to the distinct itself.

Reply:
After debugging, I got what you meant. So I wonder how to convert the `ColumnBinding` in `Distinct` to the `ColumnBinding` in the `Join` node. Are there any functions that can be referenced?

Reply:
It seems a little bit hard to convert. So just wondering, why not set the unique column index for each column? Because we can easily distinguish them through the unique column ID. 
For example, if we have two tables
```
create table t(a int, b int);   -- column_index_t_a=0, column_index_t_b=1
create table t1(a int, b int, c int);  -- column_index_t1_a=2, column_index_t1_b=3, column_index_t1_c=4
```